// ts-jest/dist/postprocess.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var logger_1 = require("./utils/logger");
var get_ts_jest_config_1 = require("./utils/get-ts-jest-config");
var imports_1 = require("./utils/imports");
var babel;
var istanbulPlugin;
var jestPreset;
function importBabelDeps() {
    if (babel)
        return;
    babel = imports_1.importBabelCore();
    istanbulPlugin = imports_1.importBabelPluginIstanbul();
    jestPreset = imports_1.importBabelPresetJest();
}
function postProcessCode(jestConfig, transformOptions, transpileOutput, filePath) {
    var postHook = exports.getPostProcessHook(jestConfig);
    return postHook(transpileOutput, filePath, jestConfig, transformOptions);
}
exports.postProcessCode = postProcessCode;
function createBabelTransformer(options) {
    importBabelDeps();
    var presets = options.presets.slice();
    var plugins = options.plugins.slice();
    if (!hasBabelAddon(presets, jestPreset)) {
        presets.push(jestPreset);
    }
    var hasIstanbul = hasBabelAddon(plugins, istanbulPlugin);
    var optionsBase = __assign({}, options, { presets: presets, plugins: plugins });
    delete optionsBase.cacheDirectory;
    var babelTransformer = function (codeSourcemapPair, filename, config, transformOptions) {
        var inputSourceMap = typeof codeSourcemapPair.map === 'string'
            ? JSON.parse(codeSourcemapPair.map)
            : codeSourcemapPair.map;
        var theseOptions = __assign({}, optionsBase, { filename: filename,
            inputSourceMap: inputSourceMap });
        if (transformOptions && transformOptions.instrument) {
            theseOptions.auxiliaryCommentBefore = ' istanbul ignore next ';
            if (!hasIstanbul) {
                theseOptions.plugins = theseOptions.plugins.concat([
                    istanbulPluginConfig(config),
                ]);
            }
        }
        return babel.transform(codeSourcemapPair.code, theseOptions);
    };
    return babelTransformer;
}
exports.getPostProcessHook = function (jestConfig) {
    var tsJestConfig = get_ts_jest_config_1.default(jestConfig);
    if (tsJestConfig.skipBabel) {
        logger_1.logOnce('Not using any postprocess hook.');
        return function (input) { return input; };
    }
    var tsJestBabelConfig = tsJestConfig.babelConfig || {};
    var babelOptions = __assign({}, tsJestBabelConfig, { babelrc: tsJestConfig.useBabelrc || false, plugins: toArray(tsJestBabelConfig.plugins), presets: toArray(tsJestBabelConfig.presets), sourceMaps: tsJestConfig.disableSourceMapSupport ? false : 'both' });
    logger_1.logOnce('Using babel with options:', babelOptions);
    return createBabelTransformer(babelOptions);
};
function toArray(iter) {
    return iter ? Array.from(iter) : [];
}
function istanbulPluginConfig(jestConfig) {
    return [
        istanbulPlugin,
        {
            cwd: jestConfig.rootDir,
            exclude: [],
        },
    ];
}
function hasBabelAddon(inputList) {
    var addonMatches = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        addonMatches[_i - 1] = arguments[_i];
    }
    return inputList.some(function (item) {
        return addonMatches.indexOf(Array.isArray(item) ? item[0] : item) !== -1;
    });
}
